fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/impl/event/lifecycle/LifecycleInternalListeners.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.event.lifecycle;
+
+import java.util.List;
+
+import net.minecraft.block.entity.BlockEntity;
+import net.minecraft.entity.Entity;
+import net.minecraft.server.world.ServerWorld;
+
+import net.fabricmc.api.ModInitializer;
+import net.fabricmc.fabric.api.event.lifecycle.v1.ServerLifecycleEvents;
+
+public class LifecycleInternalListeners implements ModInitializer {
+	@Override
+	public void onInitialize() {
+		ServerLifecycleEvents.SERVER_STOP.register(server -> {
+			// We use the server shutdown to unload all entities and block entities so their events are fired.
+			for (ServerWorld world : server.getWorlds()) {
+				final List<Entity> entities = world.getEntities(null, entity -> true); // Get every single entity in the world
+
+				for (Entity entity : entities) {
+					ServerLifecycleEvents.ENTITY_UNLOAD.invoker().onEntityUnload(entity, world);
+				}
+
+				for (BlockEntity blockEntity : world.blockEntities) {
+					ServerLifecycleEvents.BLOCK_ENTITY_UNLOAD.invoker().onUnloadBlockEntity(blockEntity, world);
+				}
+			}
+		});
```

Comment:
We might want to impose a stricter order for when the SERVER_STOP and ENTITY_UNLOAD/BLOCK_ENTITY_UNLOAD events are called. Let's say a mod registers a listener for both SERVER_STOP and ENTITY_UNLOAD/BLOCK_ENTITY_UNLOAD. Depending on whether the mod gets to register its SERVER_STOP event listener _before_ or _after_ this listener (which may be nondeterministic), it will have its SERVER_STOP event listener called either before or after all the ENTITY_UNLOAD/BLOCK_ENTITY_UNLOAD events are invoked.

I imagine a mod might use these events to free resources, and might want to use some server-bound resource which is freed when the SERVER_STOP event is called when entities and block entities are unloaded. Having all the entities and block entities unload before the server is unloaded would make the most sense.

Reply:
I could fire that internal listener a bit earlier so all entities would be ublosded before the server stops. However some event listeners may already use the fact not every entity is unloaded during server shutdown themselves, so we would be shooting them in the foot then. I could hook this into the server world being closed so it is garunteed the world no longer will exist, and some what covers custom worlds added by other mods

Reply:
@JamiesWhiteShirt So I decided to make this case be bound to the closing of the server worlds so events can still work off the old contract where entities were still present on the server